src/cryptography/x509/general_name.py

```diff
@@ -173,13 +173,34 @@ def __ne__(self, other):
 @utils.register_interface(GeneralName)
 class UniformResourceIdentifier(object):
     def __init__(self, value):
-        if not isinstance(value, six.text_type):
-            raise TypeError("value must be a unicode string")
+        if isinstance(value, six.text_type):
+            try:
+                value = value.encode("ascii")
+            except UnicodeEncodeError:
+                value = self._idna_encode(value)
+                warnings.warn(
+                    "UniformResourceIdentifier values should be passed as "
+                    "idna-encoded bytes, not strings. Support for passing "
```

Comment:
`idna-encoded` is an over simplification right? the hostname portion should be idna-encoded, but the path segment should use the urlquote stuff, right?